core/src/main/java/com/aem/demo/core/models/MainPageModel.java

```diff
@@ -65,24 +64,31 @@ private void init() {
             holdModels.add(jcrContent.adaptTo(NewsCardModel.class));
         }
 
-        double listSize = holdModels.size();
-        double listPart = listSize / 5;
-        double roundPart = Math.round(listPart);
+        String searchText = Optional.ofNullable(servletRequest).map(SlingHttpServletRequest::getRequestParameterMap)
+                .map(map -> map.getValue("search"))
+                .map(RequestParameter::getString)
+                .orElse(null);
 
-        if (roundPart != listPart) {
-            roundPart++;
-        }
 
-        pageNumbers = IntStream.range(1, (int) (roundPart + 1)).boxed().collect(Collectors.toList());
+        String paginationNumber = Optional.ofNullable(servletRequest).map(SlingHttpServletRequest::getRequestParameterMap)
+                .map(map -> map.getValue("page"))
+                .map(RequestParameter::getString)
+                .orElse(null);
+
+        isNumeric(paginationNumber);
 
-        String searchText = servletRequest.getParameter("search");
-        String paginationNumber = servletRequest.getParameter("page");
 
         if (searchText != null) {
             models = searchService.retrieveModels(searchText);
+            paginateModels(models);
```

Comment:
Please note that `paginateModels` repeats three times in every branch of `if`. This is not good